### DISCUSSION TITLE
sliding window calculations?

### DISCUSSION THREAD

**Discussion by mwotton** [TimefoldAI/timefold-solver#602] [General]
I'm trying to implement a constraint that's something like "the maximum number of hours worked in any N day period by a single employee is M" in an employee scheduling problem, where hours are a property of individual shifts. In Postgresql I'd probably use some kind of window function (PARTITION OVER etc), but I couldn't find a primitive that did that in timefold. I can sort of see how I could use toList once I group over Employee by sorting and manually sliding a time window over that list, but I was wondering: is there a more idiomatic way to handle this?

**Comment by mwotton** [comment]
(maybe toList is ok if it's the absolute last step in the constraint chain?)

**Reply by triceo** [reply]
The problem with lists (or any other collection) in constraints is that you need the list to _iterate_ over the list. And it is the iteration that doesn't scale; the longer the list, the longer the iteration.

That said, if the list is very small - single digit sizes - then I could see how using a `toList()` immediately before `penalize()` would't be a big problem.

**Comment by triceo** [comment]
Your answer most likely lies in the [consecutive constraint collector](https://timefold.ai/docs/timefold-solver/latest/constraints-and-score/score-calculation#collectorsConsecutive). This allows you to group data into clusters, and to check their length as well as the length of the gaps between the clusters. This operation is fully incremental.

**Reply by mwotton** [reply]
Thanks, that looks like the right thing
